apps/vm-hotreload/webpack-custom-hmr/src/custom-hmr-helpers.js

```diff
@@ -0,0 +1,858 @@
+const vm = require('vm');
+
+// Webpack runtime code for in-memory HMR - only executed when webpack is available
+function injectInMemoryHMRRuntime(__webpack_require__) {
+  if (typeof __webpack_require__ === 'undefined') {
+    console.warn(
+      '[Custom HMR Helper] __webpack_require__ not available, skipping runtime injection',
+    );
+    return;
+  }
+
+  // we need to patch the runtime module for chunk loading hot update chunks to support in memory
+  /******/ /* webpack/runtime/readFile chunk loading */
+  /******/ (() => {
+    /******/ // no baseURI
+    /******/
+    /******/ // object to store loaded chunks
+    /******/ // "0" means "already loaded", Promise means loading
+    /******/ var installedChunks = (__webpack_require__.hmrS_readFileVm =
+      __webpack_require__.hmrS_readFileVm || {
+        /******/ index: 0,
+        /******/
+      });
+    /******/
+    /******/ // Global storage for in-memory chunk content
+    /******/ var inMemoryChunks = {};
+    /******/
+    /******/ // no on chunks loaded
+    /******/
+    /******/ // no chunk install function needed
+    /******/
+    /******/ // no chunk loading
+    /******/
+    /******/ // no external install chunk
+    /******/
+    /******/ function loadUpdateChunk(chunkId, updatedModulesList) {
```

Comment:
node federation plugin already patches `require.f.readFileVM` but i never updated it to include the HMR stuff - in future this code could live in the node runtime plugin where other chunk loading stuff already exists.